FEA TargetEncoder should respect sample_weights

MiguelParece opened the pull request:
This commit updates the TargetEncoder to respect sample_weights. This Fixes #28881 and modifies the files: _target_encoder.py, _target_encoder_fast.pyx, and test_target_encoder.py.

DuarteSJ committed:
FEA TargetEncoder should respect sample_weights

This commit updates the TargetEncoder to respect sample_weights.
This addresses issue #28881 and modifies the files:
_target_encoder.py, _target_encoder_fast.pyx, and test_target_encoder.py.

Co-authored-by: Miguel Cabral Parece <miguelparece@tecnico.ulisboa.pt>

DuarteSJ commented:
I still have a few questions I would appreciate if someone could clarify. @mayer79 @betatim 

    Is our interpretation of the `sample_weight` correct?

    In cases where `sample_weight` is not passed to the methods:
        Should we just create an array of 1's?
        Or should we avoid the unnecessary multiplications, even if it results in generating code that is a bit less readable?

DuarteSJ committed:
Fixed Linting Issues

Co-authored-by: Miguel Cabral Parece <miguelparece@tecnico.ulisboa.pt>

DuarteSJ committed:
Fixed a bug in fast target encoder
Type was wrong for sample weights.

Co-authored-by: Miguel Cabral Parece <miguelparece@tecnico.ulisboa.pt>"

DuarteSJ committed:
Fixed code coverage problems

Co-authored-by: Miguel Cabral Parece <miguelparece@tecnico.ulisboa.pt>

DuarteSJ committed:
Added entry to the whats_new document

Co-authored-by: Miguel Cabral Parece <miguelparece@tecnico.ulisboa.pt>

DuarteSJ committed:
Fixed doc/whats_new/v1.6.rst

ogrisel commented:
> Should we just create an array of 1's?
> Or should we avoid the unnecessary multiplications, even if it results in generating code that is a bit less readable?

Do can measure the performance impact of setting them to 1s as done in this PR against the current state in `main`? I suppose it should not matter too much. If it's less than 30% I wouldn't worry too much as I doubt that `TargetEncoder` can ever become a performance bottleneck in ML pipelines.

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -396,42 +417,64 @@ def _fit_encodings_all(self, X, y):
+            # Maybe leave `sample_weight` as None so as to make
+            # _target_encoder_fast.pyx more efficient
+            if sample_weight is None:
+                sample_weight = np.ones_like(y)
```
You might want to use `sklearn.utils.validation._check_sample_weight` instead.

ogrisel commented on sklearn/preprocessing/_target_encoder.py:
```diff
@@ -396,42 +417,64 @@ def _fit_encodings_all(self, X, y):
+            # Maybe leave `sample_weight` as None so as to make
+            # _target_encoder_fast.pyx more efficient
+            if sample_weight is None:
+                sample_weight = np.ones_like(y)
```
Here again, `sklearn.utils.validation._check_sample_weight`. But this should probably be done in the public API and then make `sample_weight` a non-optional parameter of `_fit_encoding_binary_or_continuous`.

ogrisel commented on sklearn/preprocessing/tests/test_target_encoder.py:
```diff
@@ -22,32 +22,64 @@
+        if sample_weight is not None:
+            y_variance = np.sum(sample_weight * (y_numeric - y_mean) ** 2) / np.sum(
+                sample_weight
+            )
```
I think we might want to reuse `mean_variance_axis` instead of reimplementing weighted mean and variance computation in this function.

DuarteSJ replied:
Mean always needs to be computed. Variance is only computed in cases where `smooth=auto`. Should I just use `_incremental_mean_and_var` and always compute the both of them even if variance isn't going to be used?

DuarteSJ committed:
Refactored _target_encoder
When sample weights are None, innitialize them as an array of 1's
for better code readability and consistency.

Co-authored-by: Miguel Cabral Parece <miguelparece@tecnico.ulisboa.pt>

DuarteSJ commented:
@ogrisel What do you think we should do?

bhavek-jamnadas commented:
Hey, any reason this was closed?

Would love to have it implemented.

DuarteSJ commented:
> Hey, any reason this was closed?
> 
> Would love to have it implemented.

Hey there! I believe I was close to finishing it (if I recall correctly it was fully implemented and working, but I am not 100% sure since it was long ago), but the guy who had started working on it with me closed the pr after 4 months with no response.

ogrisel commented:
@DuarteSJ are you interested in reviving this PR or would you prefer someone else to takeover?

DuarteSJ commented:
> @DuarteSJ are you interested in reviving this PR or would you prefer someone else to takeover?

Yes, I'm interested in reviving this PR. I’ll take some time to review your previous message to ensure I fully understand it. If I have any questions, I’ll follow up.
Should I open a new pr with the same name? Or what would be the standard practice here?

DuarteSJ commented:
@ogrisel
